File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
  const map = new Map();
  
  const deepCopy = (node) => {
    if (!node) return null;
    if (map.has(node)) return map.get(node);
    
    const currNode = new Node(node.val);
    map.set(node, currNode);
    currNode.next = deepCopy(node.next);
    currNode.random = deepCopy(node.random);
    
    return currNode;
  }
  
  return deepCopy(head);
};

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        def deepCopy(node, seen):
            if not node:
                return None
            if node in seen:
                return seen[node]
            copy = Node(node.val)
            seen[node] = copy
            copy.next = deepCopy(node.next, seen)
            copy.random = deepCopy(node.random, seen)
            return copy
          
          
        return deepCopy(head, {})
        